frame/identity/src/lib.rs

```diff
@@ -0,0 +1,1040 @@
+			});
+
+			Self::deposit_event(RawEvent::RegistrarAdded(i));
+		}
+
+		/// Set an account's identity information and reserve the appropriate deposit.
+		///
+		/// If the account already has identity information, the deposit is taken as part payment
+		/// for the new deposit.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have a registered
+		/// identity.
+		///
+		/// - `info`: The identity information.
+		///
+		/// Emits `IdentitySet` if successful.
+		///
+		/// # <weight>
+		/// - `O(X + R)` where `X` additional-field-count (deposit-bounded).
+		/// - At most two balance operations.
+		/// - One storage mutation (codec `O(X + R)`).
+		/// - One event.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(50_000)]
+		fn set_identity(origin, info: IdentityInfo) {
+			let sender = ensure_signed(origin)?;
+			let fd = <BalanceOf<T>>::from(info.additional.len() as u32) * T::FieldDeposit::get();
+
+			let mut id = match <IdentityOf<T>>::get(&sender) {
+				Some(mut id) => {
+					// Only keep non-positive judgements.
+					id.judgements.retain(|j| j.1.is_sticky());
+					id.info = info;
+					id
+				}
+				None => Registration { info, judgements: Vec::new(), deposit: Zero::zero() },
+			};
+
+			let old_deposit = id.deposit;
+			id.deposit = T::BasicDeposit::get() + fd;
+			if id.deposit > old_deposit {
+				T::Currency::reserve(&sender, id.deposit - old_deposit)?;
+			}
+			if old_deposit > id.deposit {
+				let _ = T::Currency::unreserve(&sender, old_deposit - id.deposit);
+			}
+
+			<IdentityOf<T>>::insert(&sender, id);
+			Self::deposit_event(RawEvent::IdentitySet(sender));
+		}
+
+		/// Set the sub-accounts of the sender.
+		///
+		/// Payment: Any aggregate balance reserved by previous `set_subs` calls will be returned
+		/// and an amount `SubAccountDeposit` will be reserved for each item in `subs`.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have a registered
+		/// identity.
+		///
+		/// - `subs`: The identity's sub-accounts.
+		///
+		/// # <weight>
+		/// - `O(S)` where `S` subs-count (hard- and deposit-bounded).
+		/// - At most two balance operations.
+		/// - One storage mutation (codec `O(S)`); one storage-exists.
+		/// # </weight>
+		fn set_subs(origin, subs: Vec<(T::AccountId, Data)>) {
+			let sender = ensure_signed(origin)?;
+			ensure!(<IdentityOf<T>>::exists(&sender), "not found");
+			ensure!(subs.len() <= T::MaximumSubAccounts::get() as usize, "too many subs");
+
+			let old_deposit = <SubsOf<T>>::get(&sender).0;
+			let new_deposit = T::SubAccountDeposit::get() * <BalanceOf<T>>::from(subs.len() as u32);
+
+			if old_deposit < new_deposit {
+				T::Currency::reserve(&sender, new_deposit - old_deposit)?;
+			}
+			// do nothing if they're equal.
+			if old_deposit > new_deposit {
+				let _ = T::Currency::unreserve(&sender, old_deposit - new_deposit);
+			}
+
+			if subs.is_empty() {
+				<SubsOf<T>>::remove(&sender);
```

Comment:
What happens with the `deposit`? Should we error in case it's not unreserved (i.e. `new_deposit.is_zero()`)?

Reply:
zero subs, zero remaining deposit.